graph-commons/src/test/scala/io/renku/http/client/GitLabClientSpec.scala

```diff
@@ -0,0 +1,191 @@
+import io.renku.graph.model.GitLabUrl
+import io.renku.http.client.RestClientError.UnauthorizedException
+import io.renku.interpreters.TestLogger
+import io.renku.logging.TestExecutionTimeRecorder
+import io.renku.metrics.GitLabApiCallRecorder
+import io.renku.stubbing.ExternalServiceStubbing
+import io.renku.testtools.IOSpec
+import org.http4s.Method.{GET, _}
+import org.http4s.Status.{NotFound, Ok, Unauthorized}
+import org.http4s.circe.CirceEntityCodec.circeEntityDecoder
+import org.http4s.{Method, Request, Response, Status, Uri}
+import org.scalacheck.Gen
+import org.scalamock.scalatest.MockFactory
+import org.scalatest.matchers.should
+import org.scalatest.wordspec.AnyWordSpec
+
+class GitLabClientSpec
+    extends AnyWordSpec
+    with IOSpec
+    with ExternalServiceStubbing
+    with should.Matchers
+    with MockFactory {
+
+  "send" should {
+
+    "fetch json from a given page" in new TestCase {
+
+      stubFor {
+        callMethod(method, s"api/v4/$path")
+          .willReturn(
+            okJson(result)
+              .withHeader("X-Next-Page", maybeNextPage.map(_.show).getOrElse(""))
+          )
+      }
+
+      client.send(method, path, endpointName)(_ => IO(result)).unsafeRunSync() shouldBe result
+    }
+
+    "fetch commits using the given personal access token" in new TestCase {
+      val personalAccessToken = personalAccessTokens.generateOne
+
+      stubFor {
+        callMethod(method, s"api/v4/$path")
+          .withAccessToken(personalAccessToken.some)
+          .willReturn(
+            okJson(result)
+              .withHeader("X-Next-Page", maybeNextPage.map(_.show).getOrElse(""))
+          )
+      }
+
+      client.send(method, path, endpointName)(_ => IO(result)).unsafeRunSync() shouldBe result
+    }
+
+    "fetch commits using the given oauth token" in new TestCase {
+      val authAccessToken = oauthAccessTokens.generateOne
+
+      stubFor {
+        callMethod(method, s"api/v4/$path")
+          .withHeader("PRIVATE-TOKEN", equalTo(authAccessToken.value))
+          .willReturn(
+            okJson(result)
+              .withHeader("X-Next-Page", maybeNextPage.map(_.show).getOrElse(""))
+          )
+      }
+
+      client.send(method, path, endpointName)(_ => IO(result)).unsafeRunSync() shouldBe result
+    }
+
+    "Handle empty JSON" in new TestCase {
+      override val result: String = "{}"
+      stubFor {
+        callMethod(method, s"api/v4/$path")
+          .willReturn(
+            okJson(result)
+              .withHeader("X-Next-Page", maybeNextPage.map(_.show).getOrElse(""))
+          )
+      }
+
+      client.send(method, path, endpointName)(_ => IO(result)).unsafeRunSync() shouldBe result
+
+    }
+
+    "Handle 404 NOT FOUND without throwing" in new TestCase {
+      stubFor {
+        callMethod(method, s"api/v4/$path")
+          .willReturn(
+            notFound()
+          )
+      }
+
+      client.send(method, path, endpointName)(_ => IO(())).unsafeRunSync() shouldBe ()
```

Comment:
What if we make the mapping from lines 114, 133 and 149 global and use it in all the tests?